csharp/src/AddressBook/Addressbook.cs

```diff
@@ -76,34 +76,60 @@ public sealed partial class Person : pb::IMessage<Person>, global::System.IEquat
       get { return global::Google.ProtocolBuffers.Examples.AddressBook.Addressbook.internal__static_tutorial_Person__FieldAccessorTable; }
     }
 
+    private bool _frozen = false;
+    public bool IsFrozen { get { return _frozen; } }
+
     public Person() { }
+
     public Person(Person other) {
-      MergeFrom(other);
+      name_ = other.name_;
+      id_ = other.id_;
+      email_ = other.email_;
+      phone_ = other.phone_.Clone();
+    }
+
+    public Person Clone() {
+      return new Person(this);
+    }
+
+    public void Freeze() {
+      if (IsFrozen) {
+        return;
+      }
+      _frozen = true;
+      phone_.Freeze();
     }
+
     public const int NameFieldNumber = 1;
     private string name_ = "";
     public string Name {
       get { return name_; }
-      set { name_ = value ?? ""; }
+      set {
+        pb::Freezable.CheckMutable(this);
```

Comment:
i think inlining like this
if (_frozen) {
   pb:Freezable.Throw()
}
would allow JIT to optimize out the static call entirely, but I'm just guessing.